src/mtype.d

```diff
@@ -6199,10 +6204,19 @@ extern (C++) final class TypeFunction : TypeNext
                     fparam.storageClass |= STCreturn;
                 }
 
-                if (fparam.storageClass & STCreturn && !(fparam.storageClass & (STCref | STCout)))
+                if (fparam.storageClass & STCreturn)
                 {
-                    error(loc, "'return' can only be used with 'ref' or 'out'");
-                    errors = true;
+                    if (!(fparam.storageClass & (STCref | STCout)))
+                    {
+                        error(loc, "'return' can only be used with 'ref' or 'out'");
+                        errors = true;
+                    }
+                    else if (!tf.isref && tf.next && !tf.next.hasPointers())
+                    {
+                        error(loc, "parameter %s is 'return' but function does not return any indirections",
```

Comment:
How about `parameter %s is 'return' but it's not returned by reference`? That way the message gives insight about the purpose of the attribute. Another option is: `parameter %s should be 'return' only if it is returned by reference`.
I think it's safe to say that if the function doesn't return any indirections that it's not possible to return any parameter by reference/address.
BTW apart from writing a test about this case, we may also need a test where there are two `return` parameters and only one of them is returned *unconditionally*.